### DISCUSSION TITLE
How to use IndentedBlock correctly

### DISCUSSION THREAD

**Discussion by chaihub** [pyparsing/pyparsing#430] [Q&A] [answered]
Greetings!

I am using version 3.0.9, and am trying to use `IndentedBlock()` to parse this simple directory + file listing. I am not getting the expected results. A line that is not indented is being matched successfully by this function. I want to find out how to use it correctly in this scenario. Any suggestions? Thanks in advance!

Here is my code:
```
from pyparsing import *

DATA1 = """
Directory1
    Filea
    Fileb
    Filec
Directory2
Directory3
"""

DATA2 = """
Directory1
Directory2
    Filea
    Fileb
    Filec
Directory3
"""

dirData = Word(alphanums)
fileData = Word(alphas)
dirListing = dirData + IndentedBlock(fileData)

parseTree = dirListing.parse_string(DATA1)
parseTree.pprint()
parseTree = dirListing.parse_string(DATA2)
parseTree.pprint()
```

And here is the output (the first line shows how I want the output; the second line shows the undesirable output where a non-indented word has matched successfully):
```
['Directory1', ['Filea', 'Fileb', 'Filec']]
['Directory1', ['Directory']]
```

**Comment by thatrandomperson5** [comment, marked as answer]
This is a late response, but if you are still interested, here is the sultion:
```py
from pyparsing import *

DATA1 = """
Directory1
    Filea
    Fileb
    Filec
Directory2
Directory3
"""

DATA2 = """
Directory1
Directory2
    Filea
    Fileb
    Filec
Directory3
"""

dirData = Word(alphanums)
fileData = Word(alphas)
dirListing = OneOrMore(dirData+(IndentedBlock(fileData) | Suppress(StringEnd())))

parseTree = dirListing.parse_string(DATA1)
parseTree.pprint()
parseTree = dirListing.parse_string(DATA2)
parseTree.pprint()
```
I added an indent block match-first with a suppressed stringEnd to make sure that there were no glitches about where lines ended ( I got some of those). I also added one or more so that your code could actually parse all lines. Why your code did not work was because A: it was forcing pyparsing to find and indent block even if it was not there, B: it only parsed one directory.

**Warning:** either my code is not completly right or there is a bug with pyparsing but it seems to be misinterpreting things as indented. This is meant to help you find your answer, not be your answer, so if you get this issue ill leave you to fix it.

Hope this helps!

**Reply by chaihub** [reply]
Thank you very much for your response, @thatrandomperson5 ! It clarified some points for me and helped me experiment further.

Also, upon re-reading the documentation, I realized I had missed this note: "... a valid IndentedBlock must contain at least 1 matching statement_expr ..." . I take that to mean that, in my examples, I must have at least one file listed under each directory for IndentedBlock to work correctly. So I should find some other way if empty directories are to be detected correctly. I shall try some other ways.

Cheers!
